web/packages/teleterm/src/mainProcess/protocolHandler.ts

```diff
@@ -36,45 +38,135 @@ const disabledSchemes = [
   'gopher',
   'javascript',
   'mailto',
+  'file',
 ];
+const APP_FILE_SCHEMA = 'app-file';
+const CONNECT_AUTHORITY = 'connect-app';
+
+/**
+ * Builds a URI for the custom 'app-file://' scheme.
+ * The path will be resolved against `app.getAppPath()`.
+ */
+export function buildAppFileUri(relativePath: string) {
+  const baseUrl = `${APP_FILE_SCHEMA}://${CONNECT_AUTHORITY}`;
+
+  const uri = new URL(relativePath, baseUrl);
+  return uri.toString();
+}
 
-// these protocols are not used within the app
+/** Disables protocols not used by the app. */
 function disableUnusedProtocols() {
   disabledSchemes.forEach(scheme => {
-    protocol.interceptFileProtocol(scheme, (_request, callback) => {
+    protocol.handle(scheme, () => {
+      const message = `Denying request: Invalid scheme (${scheme})`;
       logger.error(`Denying request: Invalid scheme (${scheme})`);
-      callback({ error: -3 });
+      return new Response(message, {
+        status: 403,
+        statusText: 'Forbidden Protocol',
+      });
     });
   });
 }
 
-// intercept, clean, and validate the requested file path.
-function interceptFileProtocol() {
-  const installPath = app.getAppPath();
+/**
+ * Registers the 'http://' protocol handler.
+ * Adds cross-origin headers to handled responses to enable features requiring
+ * cross-origin isolation.
+ */
+function handleHttpProtocol(): void {
+  protocol.handle('http', async request => {
+    const response = await net.fetch(request, {
+      // Must be true to prevent the handler from calling itself and entering an infinite loop.
+      bypassCustomProtocolHandlers: true,
+    });
+    setCrossOriginIsolationHeaders(response.headers);
+    return response;
+  });
+}
 
-  protocol.interceptFileProtocol('file', (request, callback) => {
-    const target = path.normalize(fileURLToPath(request.url));
-    const realPath = fs.existsSync(target) ? fs.realpathSync(target) : target;
+/**
+ * Registers the 'app-file://' protocol handler.
+ * Serves application files from the build directory and adds
+ * cross-origin headers to responses, enabling features that
+ * require cross-origin isolation.
+ */
+function handleAppFileProtocol(): void {
+  const appPath = app.getAppPath();
 
-    if (!path.isAbsolute(realPath)) {
-      logger.error(`Denying request to non-absolute path '${realPath}'`);
-      return callback({ error: -3 });
+  protocol.handle(APP_FILE_SCHEMA, async request => {
+    const filePath = decodeURIComponent(new URL(request.url).pathname).replace(
+      // Remove the leading slash.
+      /^\/+/,
+      ''
+    );
+    const target = path.join(appPath, filePath);
+    let realPath: string;
+    try {
+      realPath = await fs.promises.realpath(target);
```

Comment:
I simplified the logic a bit to avoid two sync calls (`fs.existsSync`,  `fs.realpathSync`) for each file.
But as a result, now we will return 400 also if the file doesn't exist.